### ISSUE TITLE
Feature request: could `--version` print the full `Base.VERSION`?

### ISSUE THREAD

**Issue by DilumAluthge** [JuliaLang/julia#39167]
```bash
$ julia --version
julia version 1.7.0-DEV
$ julia -e 'println("julia version ", Base.VERSION)'
julia version 1.7.0-DEV.243
```

It would be nice if `julia --version` printed the output `julia version 1.7.0-DEV.243`.

cc: @KristofferC

**Comment by StefanKarpinski** [comment]
This would be a nice feature if someone wanted to implement this.

**Comment by TheCedarPrince** [comment]
I was just looking through base on how to fix this.
Do I want to modify the `src/Makefile` and `src/jloptions.c` file to produce the full path?
Or where should one look?
Thanks!

**Comment by Moelf** [comment]
not that easy, to ensure 100% same output you have to replicate this logic inside `jloptions.c`
https://github.com/JuliaLang/julia/blob/dbae5a50e506f05a4d4f2f3471c76291e5252df4/base/version.jl#L236-L258

**Comment by vmpyr** [comment]
Hi, I want to take up this issue if no one is working on it. Also, I was wondering can not just simply run `julia -e 'println("julia version ", Base.VERSION)'` when `--version` flag is called?

**Comment by vmpyr** [comment]
After some digging I've found that we would have to change this bit of code in the Makefile:
```
$(BUILDDIR)/julia_version.h: $(JULIAHOME)/VERSION
	@echo "// This is an autogenerated header file" > $@.$(JULIA_BUILD_MODE).tmp
	@echo "#ifndef JL_VERSION_H" >> $@.$(JULIA_BUILD_MODE).tmp
	@echo "#define JL_VERSION_H" >> $@.$(JULIA_BUILD_MODE).tmp
	@echo "#define JULIA_VERSION_STRING" \"$(JULIA_VERSION)\" >> $@.$(JULIA_BUILD_MODE).tmp
	@echo $(JULIA_VERSION) | awk 'BEGIN {FS="[.,-]"} \
	{print "#define JULIA_VERSION_MAJOR " $$1 "\n" \
	"#define JULIA_VERSION_MINOR " $$2 "\n" \
	"#define JULIA_VERSION_PATCH " $$3 ; \
	if (NF<4) print "#define JULIA_VERSION_IS_RELEASE 1" ; else print  "#define JULIA_VERSION_IS_RELEASE 0"}' >> $@.$(JULIA_BUILD_MODE).tmp
	@echo "#endif" >> $@.$(JULIA_BUILD_MODE).tmp
	mv $@.$(JULIA_BUILD_MODE).tmp $@
```

The `JULIA_VERSION_STRING` variable doesn't contain the full version name along with build number. I think this would be pretty easy to do. I can try doing it.

**Comment by vmpyr** [comment]
Hi, so I tried doing this by editing the `jloptions.c` file but it doesn't work as expected and I'm getting a `segmentation fault (core dumped)` error when I try `julia -v`.

Here are the changes I made:
```
        case 'v': // version
            jl_function_t *string_func = jl_get_function(jl_base_module, "string");
            jl_value_t *vrs_vrsnum = jl_get_function(jl_base_module, "VERSION");
            jl_value_t *vrs_num = jl_call1(string_func, vrs_vrsnum);
            jl_printf(JL_STDOUT, "julia version %s\n", jl_string_data(vrs_num));
            exit(0);
```

Where am I going wrong?

**Comment by StefanKarpinski** [comment]
The context that this is what you were trying to do was absent on Slack. There's no Julia runtime to evaluate code with when you're parsing Julia's options. How could there be? You have to parse the options to know how to start the runtime in the first place.

One non-circular option is to insert the build number into the `src/version.h` that is already generated by makefiles. However, that's a big pain because the build number changes every time any commit is made, and modifying the `src/version.h` file will cause all C code to need to be recompiled. So doing that would mean that any time a git commit is made all C code needs to be recompiled.

**Comment by vmpyr** [comment]
That makes a lot of sense. What I thought initially was the julia code would've been ready since it is already being used by functions like `jl_printf` but yes, that isn't actually julia code but C code.

How can we do this then, a change to `Makefile` was what I was thinking about early on, because that is what generates the header file that contains the version number. But I don't understand the syntax or anything about Makefiles to be honest. Is going the `Makefile` approach correct?

**Comment by inkydragon** [comment]
For reference: `GIT_VERSION_INFO` is defined here.
https://github.com/JuliaLang/julia/blob/17dee39962e8aed2b2c3146819de8b366bcb98f8/base/version_git.sh#L104-L115

> Is going the Makefile approach correct?

Modifying the Makefile is a possible approach,
but I'm not sure that the benefits of doing so outweigh the troubles it causes.

----

Considering this approach, you may need to understand and modify the build process.
I don't think this issue is a `good first issue` anymore.

**Comment by vmpyr** [comment]
>Modifying the Makefile is a possible approach

What are the other possible ways to do this then? If only `Makefile` changes can do it, then I'll have to first learn and understand those, will be a plus anyway.

**Comment by vtjnash** [comment]
Yes, we currently generate that file into base/version_git.jl base/build_h.jl and src/julia_version.h from the Makefile, so you could update the content of julia_version.h to have more info (as generated into the other files), and update the printing to print more of the data

**Comment by vmpyr** [comment]
Yes, but `src/julia_version.h` is a generated file right? So it's content can only be altered my making changes to `src/Makefile`?

**Comment by vtjnash** [comment]
Yes
